streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java

```diff
@@ -1140,6 +1141,24 @@
                                      final ValueJoiner<? super V, ? super VO, ? extends VR> joiner,
                                      final Materialized<K, VR, KeyValueStore<Bytes, byte[]>> materialized);
 
+    /**
+     *
+     * A many:1 join with the other table. The foreignKeyExtractor selects the key to join with the other table.
+     *
+     * @param other the table containing the records to be joined on. Keyed by KO
+     * @param foreignKeyExtractor extracts the key (KO) from this table's value (V)
+     * @param joiner specifies how to join the records from both tables
+     * @param materialized the materialized output store
+     * @param <VR> the value type of the result {@code KTable}
+     * @param <KO> the key type of the other {@code KTable}
+     * @param <VO> the value type of the other {@code KTable}
+     * @return
+     */
+    <VR, KO, VO> KTable<K, VR> join(final KTable<KO, VO> other,
+                                    final ValueMapper<V, KO> foreignKeyExtractor,
```

Comment:
To me it felt natural as I wanted to apply a mapping function to a value and obtain a result. Semantically they feel the same though, so I don't really care one way or another. If you think it's more appropriate to go with `java.util.function.Function<V, KO>` I will make the change.

Reply:
Hey @bellemare, I think your feeling is reasonable, but to me it actually supports using a Function more than using a ValueMapper. What I mean is, you just want to run a _function_ on the value and obtain a result. The ValueMapper implies you're mapping the value to a new value to produce a new KTable with the result value. In other words, it only makes sense in a `mapValues()` operation.

The ValueMapper interface was introduced for the `mapValues()` operator, and we abused (along with others of our functional interfaces) by using it other places where we just want a unary function (because Java had no stdlib functions at that point).

Now that Java has Function, we can clean this mess up. Right now, though, I'm just proposing not to create more mess that we have to clean up later. If people disagree with using Function (for some reason), then we should make a new semantically correct interface instead of using ValueMapper.

Reply:
Function it is. Your perspective on the misuse of ValueMapper makes sense.